### DISCUSSION TITLE
How to do locale like tm-in.ts?

### DISCUSSION THREAD

**Discussion by pvanthony** [romcal/romcal#345] [Q&A] [answered]
We are trying to make the Indian Tamil language locale.

For now, we have named the file tm.ts and it seems to work with the following settings.

In index.html.
```html
    <script src="./packages/romcal/iife/romcal.js"></script>
    <script src="./packages/romcal/bundles/singapore/iife/tm.js"></script>
```
In the calendar setting.
```ts
// create object romcal
const romcal = new Romcal({
    localizedCalendar: Singapore_Tm,
    scope: 'gregorian',
    epiphanyOnSunday: true,
    corpusChristiOnSunday: true,
    ascensionOnSunday: false,
    outputOptionalMemorials: false,
    strictMode: true
});
```
It is working. Of course there are some errors because english words still appears and that is because we were not careful. We need to do more testing.

My question is how to do tm-in.ts. Because Tamil is a language from India. India has many languages so the file should be tm-in.ts.
But how to set it up?

Here is the draft of tm.ts that we are using for testing.
```ts
import { Locale } from '../types/locale';

export const locale: Locale = {
  key: 'tm',

  seasons: {
    advent: {
      season: 'திருவருகைக்காலம்',
      weekday: '$t(weekdays:{{dow}}, capitalize) of the $t(ordinals:{{week}}) திருவருகைக்கால வாரநாள்',
      sunday: '$t(ordinals:{{week}}, capitalize) திருவருகைக்கால ஞாயிறு',
      privileged_weekday: '$t(months:11, capitalize) {{day}}',
    },

    christmas_time: {
      season: 'கிறிஸ்து பிறப்பு காலம்',
      day: '$t(weekdays:{{dow}}, capitalize) கிறிஸ்து பிறப்பு காலம்',
      octave: '{{count}}º கிறிஸ்து பிறப்பின் எண்கிழமை',
      second_sunday_after_christmas: 'Second Sunday after the Nativity of the Lord',
      before_epiphany: '$t(months:0, capitalize) {{day}}',
      after_epiphany: '$t(weekdays:{{dow}}, capitalize) திருக்காட்சிக்கு பின்',
    },

    ordinary_time: {
      season: 'பொதுக்காலம்',
      weekday: '$t(weekdays:{{dow}}, capitalize) of the $t(ordinals:{{week}}) பொதுக்கால வாரநாள்',
      sunday: '$t(ordinals:{{week}}, capitalize) பொதுக்கால ஞாயிறு',
    },

    lent: {
      season: 'தவக்காலம்',
      weekday: '$t(weekdays:{{dow}}, capitalize) of the $t(ordinals:{{week}}) தவக்கால வாரநாள்',
      sunday: '$t(ordinals:{{week}}, capitalize) தவக்கால ஞாயிறு',
      day_after_ash_wed: '$t(weekdays:{{dow}}, capitalize) விபூதி புதனுக்குப் பின்',
      holy_week_day: '$t(weekdays:{{dow}}, capitalize) of புனித வாரம்',
    },

    paschal_triduum: {
      season: 'பாஸ்காவின் மூன்று புனித நாட்கள்',
    },

    easter_time: {
      season: 'பாஸ்காக்காலம்',
      weekday: '$t(weekdays:{{dow}}, capitalize) of the $t(ordinals:{{week}}) பாஸ்காக்காலம் வாரநாள்',
      sunday: '$t(ordinals:{{week}}, capitalize) பாஸ்காகாலம் ஞாயிறு',
      octave: '$t(weekdays:{{dow}}, capitalize) பாஸ்கா எண்கிழமை',
    },
  },

  periods: {
    epiphany: 'திருக்காட்சி',
    holy_week: 'புனித வாரம்',
  },

  ranks: {
    solemnity: 'பெருவிழா',
    sunday: 'ஞாயிறு',
    feast: 'திருவிழா',
    memorial: 'நினைவு',
    weekday: 'வாரநாள்',
  },

  weekdays: {
    0: 'ஞாயிறு',
    1: 'திங்கள்',
    2: 'செவ்வாய்',
    3: 'புதன்',
    4: 'வியாழன்',
    5: 'வெள்ளி',
    6: 'சனி',
  },

  months: {
    0: 'சனவரி',
    1: 'பிப்ரவரி',
    2: 'மார்ச்',
    3: 'ஏப்ரல்',
    4: 'மே',
    5: 'ஜூன்',
    6: 'ஜூலை',
    7: 'ஆகஸ்ட்',
    8: 'செப்டம்பர்',
    9: 'அக்டோபர்',
    10: 'நவம்பர்',
    11: 'டிசம்பர்',
  },

  colors: {
    black: 'கறுப்பு',
    gold: 'தங்கம்',
    green: 'பச்சை',
    purple: 'ஊதா',
    red: 'சிவப்பு',
    rose: 'ரோஸ',
    white: 'வெள்ளை',
  },

  ordinals: {
    '1': 'முதலாம்',
    '2': 'இரண்டாம்',
    '3': 'மூன்றாம்',
    '4': 'நான்காம்',
    '5': 'ஐந்தாம்',
    '6': 'ஆறாம்',
    '7': 'எழாம்',
    '8': 'எட்டாம்',
    '9': 'ஒன்பதாம்',
    '10': 'பத்தாம்',
    '11': 'பதினொன்றாம்',
    '12': 'பனிரெண்டாம்',
    '13': 'பதின்மூன்றாம்',
    '14': 'பதினான்காம்',
    '15': 'பதினைந்தாம்',
    '16': 'பதினாறாம்',
    '17': 'பதினேழாம்',
    '18': 'பதினெட்டாம்',
    '19': 'பத்தொன்பதாம்',
    '20': 'இருபதாம்',
    '21': 'இருபத்தோரம்',
    '22': 'இருபத்தி இண்டாம்',
    '23': 'இருபத்தி மூன்றாம்',
    '24': 'இருபத்தி நான்காம்',
    '25': 'இருபத்தி ஐந்தாம்',
    '26': 'இருபத்தி ஆறாம்',
    '27': 'இருபத்தி ஏழாம்',
    '28': 'இருபத்தி எட்டாம்',
    '29': 'இருபத்தி ஒன்பதாம்',
    '30': 'முப்பதாம்',
    '31': 'முப்பத்தி ஒன்றாம்',
    '32': 'முப்பத்தி இரண்டாம்',
    '33': 'முப்பத்தி மூன்றாம்',
    '34': 'முப்பத்தி நான்காம்',
  },
```

**Comment by tukusejssirs** [comment, marked as answer]
> My question is how to do tm-in.ts. Because Tamil is a language from India. India has many languages so the file should be tm-in.ts.

I am not sure if I understand your question, but I you want to know how to name the locales (`tm.ts` vs `tm-in.ts` etc). @emagnier might want to clarify this, but AFAIK the following rules apply:

1. Use [ISO 639-1 two-letter locale codes](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes). Therefore `tm.ts` → `ta.ts`.
2. Use [ISO 3166-1 alpha-2 two-letter country codes](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2).
3. Country codes should be omited when:
   - there a single country code would be used with a particular country (e.g. `sk.ts` instead of `sk-sk.ts`):
      - one exception: when a locale has different grammar rules in different parts of the worlds (e.g. `pt-br.ts`)
   - even if we have multiple locale ‘versions’ (read regional locales), but the differences are subtle (e.g. Englishes), we use one (e.g. the most used) as the main locale (i.e. without regional/country suffix; e.g. `en.ts` instead of `en-us.ts`) and other (sub-) locales would have a suffix (e.g. `en-ie.ts`, `en-gb.ts`, `en-ca.ts`), which will _override_ some keys in the main locale; another reason to keep the the country code is when in that particular country, they use different localisation string for a particular key in the localisation files;
3. Filename format is either `{locale}.ts` or `{locale}-{country}.ts`.

As for the locale code, use should change `tm.ts` to `ta.ts`.

As for the country code (should you use it or not): are there any country locale variants in different countries (like one in India and one in Singapore)? If yes, are they grammatically/linguistically _that_ different from each other? If yes, use regional code, otherwise use a single locale file (`ta.ts`).

**Reply by emagnier** [reply]
Use ISO 639-1 as the name for your localization file. 

Append the country code as ISO 3166-1 alpha-2 (separated by a dash) only if relevant: if a specific translation should only occur in this country. E.g.: `en-gb`.

**Reply by tukusejssirs** [reply]
@pvanthony, I’ve updated my previous comment with the l10n file naming rules with one note basically: another reason to keep the the country code is when in that particular country, they use different localisation string for a particular key in the localisation files. For example, if two countries, they celebration the same celebration and they use same locale (language), but name that celebration differently, you can override the celebration name in the main locale (without regional suffix) in the locale with regional locale. See `corpus_christi` in [`en.ts`](https://github.com/emagnier/romcal/blob/v3/lib/locales/en.ts#L143) and [`en-gb.ts`](https://github.com/emagnier/romcal/blob/v3/lib/locales/en-gb.ts#L22).

**Reply by pvanthony** [reply]
Thank you for the clarification. I understand better now.

**Comment by pvanthony** [comment]
Thank you very much for the reply and advice. My bad for using tm.ts instead of ta.ts. Thank you for the correction.
You are correct there is no gramatical changes that I know of. So ta.ts will be used.

**Reply by tukusejssirs** [reply]
As for renaming `tm` to `ta`: @emagnier has the last word on this. We have no rules on this gathered (yet). Also, there are multiple ISO standards that deal with country and locale codes.

**Reply by emagnier** [reply]
`ta.ts` is the way to go. 

Thank you very much for your contribution, it's very appreciated!

**Reply by tukusejssirs** [reply]
@emagnier, could you cofirm the locale and country code standards (an the rules) to use please (see [this comment of mine](https://github.com/romcal/romcal/discussions/345#discussioncomment-1836702))? Later, we could place them in the Wiki or the docs.
